code_generation/optimizer.py

```diff
@@ -0,0 +1,292 @@
+        the outputs of preceeding producers. If the scope is not global,
+        all outputs from producers in the global scope are also considered.
+    3. If the ordering is correct, return.
+    4. If the ordering is not correct,
+        4.1. find all inputs, that have to be produced before the wrong producer
+        4.2. put one producer, which is responsible for creating the input, in front of the wrong producer
+        4.3. repeat from step 2
+
+    The sorting algorithm should take at most 2*(number of producers) steps.
+    If this limit is reached, the optimization is
+    considered to be failed and an Exception is raised.
+    If a missing input cant be found in all outputs,
+    the Optimize function will raise an Exception.
+    """
+
+    def Optimize(self):
+        # first bring filters to the top
+        self.MoveFiltersUp()
+        # run optimization in a while loop
+        counter = 0
+        while not self.optimized:
+            if counter > 2 * self.size + 1:
+                log.error("Could not optimize ordering")
+                log.error("Please check, if all needed producers are activated")
+                raise Exception
+            wrongProducer, wrong_inputs = self.check_ordering()
+            producers_to_relocate = self.find_inputs(wrongProducer, wrong_inputs)
+            if len(producers_to_relocate) == 0:
+                self.optimized = True
+            else:
+                for producer_to_relocate in producers_to_relocate:
+                    counter += 1
+                    producerIndex = self.get_position(wrongProducer)
+                    self.relocate_producer(
+                        producer_to_relocate,
+                        self.get_position(producer_to_relocate),
+                        producerIndex,
+                    )
+        log.info(
+            "Optimization for scope {} done after {} steps".format(self.scope, counter)
+        )
+
+    """
+    Function used to check the ordering.
+    If at least of one the inputs of a producer cannot be found in
+    the list of all preceding outputs, the ordering is not correct.
+    If the whole odering is correct, the optimized flag is set to
+    true and the ordering is considered to be correct.
+
+    Args:
+        None
+    Returns:
+        A tuple of the wrong producer and a list of the inputs
+        that are not found in the outputs
+    """
+
+    def check_ordering(self):
+        outputs = []
+        if self.scope != "global":
+            outputs = self.global_outputs
+        for producer in self.ordering:
+            temp_outputs = producer.get_outputs(self.scope)
+            if temp_outputs != None:
+                outputs.extend(
+                    [
+                        quantity
+                        for quantity in temp_outputs
+                        if not isinstance(quantity, NanoAODQuantity)
+                    ]
+                )
```

Comment:
also see comment above on the NanoAODQuantitiy check

Reply:
you mean move the output extension after the check ? It has to be done before the check, since for producerGroups, it can very well be the case, that outputs and inputs are created during a single group. 

The optimization is only done on the level of what is provided in the config, so if a producer group is set here, the structure within the group itself is not checked.

Reply:
ok, good point!